content/posts/canvas-renderer-matplotlib-in-pyodide.md

```diff
@@ -0,0 +1,169 @@
+- [draw_markers](https://github.com/matplotlib/matplotlib/blob/9221a556d9f819b8d4c26f4fc2eca32ca67c3607/lib/matplotlib/backend_bases.py#L173): for drawing axis ticks\
+and others...
+
+In addition to the above, styles and various properties such as *join-style*, *cap-style*, *line-width*, *dashes*, etc also need to be set.
+
+Recall that [Pyodide](https://hacks.mozilla.org/2019/04/pyodide-bringing-the-scientific-python-stack-to-the-browser/) supports seamless data sharing with JavaScript and also gives the user the whole DOM available on the Python side. This essentially enables us to call functions of the *\<canvas\>* element inside a python script -- the one that defines the new backend.
+
+Thus, a simple *plt.plot()* invocation boils down to a *draw_path()* invocation for the backend which further calls the functions of the *\<canvas\>* element (responsible for actual rendering) -- all inside Python!
+
+### **Challenges**
+
+1. **Rendering Text**:
+
+The matplotlib library provides functionality for looking up fonts that it ships with. Unfortunately, these fonts are not accessible to the web browser to draw onto a canvas. We need to provide a way to load custom fonts that the browser can use.
+Thus, a simple assignment like the one below doesn’t work. 
+
+```
+from js import document
+canvas_element = document.getElementById("canvas")
+ctx = canvas_element.getContext("2d")
+ctx.font = “30px cmr10”
+```
+
+The [cmr10](https://github.com/matplotlib/matplotlib/blob/main/lib/matplotlib/mpl-data/fonts/ttf/cmr10.ttf) font is available inside matplotlib’s virtual file system but is not accessible to the browser and thus, the *\<canvas\>* element cannot use it and falls back to using a default placeholder font.
+
+This problem is solved using [Web Fonts](https://developer.mozilla.org/en-US/docs/Web/API/FontFace) which allows us to dynamically load fonts with JavaScript and make it available to the browser. Once the *“correct”* font is loaded, we can request matplotlib to redraw the plot with the correct font. Again, all of this can be done inside Python since Pyodide can interact with browser APIs and other JavaScript libraries at a very fine level.
+
+However, as easy as it sounds, keeping track of when the font is loaded (so as to request a redraw) is not so trivial. The FontFace API is asynchronous and relies on the network -- making things harder due to the following problems:
+
+-  matplotlib can call *draw_text()* multiple times in a single drawing 
+    instance -- each of them requesting the *“correct”* font to be loaded. It’s 
+    possible that a previous request for dynamically loading the correct 
+    font isn’t completed yet (still fetching from the network), and a new 
+    request has been made at that time. This creates the problem of
+    multiple font loading due to queued and concurrent invocations of the      
+    FontFace API and affects performance.
```

Comment:
I think it would probably be a better design to force people to preload the fonts they are going to use with an async call. It reminds me a bit of the Dolphin ubershaders:
https://dolphin-emu.org/blog/2017/07/30/ubershaders/

Reply:
I have replaced `dynamically` with `asynchronously`. This takes care of using the correct terms.